In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
import datetime as datetime
import warnings
warnings.filterwarnings('ignore')
import pandas 

In [2]:
contents=pd.read_csv('../orga2_21/contents.csv',low_memory=False)

In [3]:
languages=pd.read_csv('../orga2_21/languages.csv', low_memory=False)

*** 
***
 # Ejercicio 49
 #### Si decimos que un usuario sabe un idioma cuando tiene un nivel de babel mayor o
 ####            igual a 1, para aquellos que editaron una de las versiones actuales del contenido,
 ####           ¿Cuál es la tasa de revisiones sin comentario que realizan en función de los idiomas que saben?
***
***

* ___De contents nos quedamos con las columnas "revisor_id" y  "revisor_comment"___
* ___De languages filtramos por "babel_level" que sea mayor o igual a 1___

In [4]:
comentario=contents[["revisor_id","revisor_comment"]]
comentario.head(10)

revisor_id                                    revisor_comment
0   1708233.0                                     Así esta mejor
1         NaN                                   Añadir contenido
2   4642283.0  Reemplazos con [[Usuario:Benjavalero/Replacer|...
3   5338508.0  Reemplazos con [[Usuario:Benjavalero/Replacer|...
4   3615366.0  Correcciones ortográficas con [[Usuario:Benjav...
5    689866.0  Correcciones ortográficas con [[Usuario:Benjav...
6   4980693.0  Revertidos los cambios de [[Special:Contributi...
7   6331749.0                                                NaN
8   6048511.0  Correcciones ortográficas con [[Usuario:Benjav...
9    936194.0  Revertidos los cambios de [[Special:Contributi...

In [5]:
comentario.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4132164 entries, 0 to 4132163
Data columns (total 2 columns):
 #   Column           Dtype  
---  ------           -----  
 0   revisor_id       float64
 1   revisor_comment  object 
dtypes: float64(1), object(1)
memory usage: 63.1+ MB


In [6]:
comentario.isnull().sum()

revisor_id         193287
revisor_comment    469224
dtype: int64

### Como podemos ver tenemos 193287 revisor_id nulos de los cuales no nos aportara mucho a la pregunta ademas es minima a comparacion de la cantidad total de id's distintos de nulo que tenemos.

In [7]:
comentario.dropna(subset=['revisor_id'], inplace=True)
comentario.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 3938877 entries, 0 to 4132163
Data columns (total 2 columns):
 #   Column           Dtype  
---  ------           -----  
 0   revisor_id       float64
 1   revisor_comment  object 
dtypes: float64(1), object(1)
memory usage: 90.2+ MB


In [30]:
comentarionulos=comentario[comentario['revisor_comment'].isnull()]
comentarionulos.head()

revisor_id revisor_comment
7    6331749.0             NaN
21   6318667.0             NaN
22   1122788.0             NaN
24         NaN             NaN
26   5840139.0             NaN

* ___Podemos ver que revisor_id y babel_user no llevan el mismo tipo de valor por lo tanto:___

In [8]:
comentario['revisor_id']=(comentario['revisor_id']).astype(np.int64)
comentario.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3938877 entries, 0 to 4132163
Data columns (total 2 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   revisor_id       int64 
 1   revisor_comment  object
dtypes: int64(1), object(1)
memory usage: 90.2+ MB


### Creamos una columna colocando a 0 o 1 si es null o no respectivamente.

In [12]:
def cantComment_Revisor(row):
    if pd.isnull(row[1]):
        return 0
    else:
        return 1

comentario['cantComnent_Revisor']=comentario.apply(cantComment_Revisor, axis=1)
comentario.head()

revisor_id                                    revisor_comment  \
0     1708233                                     Así esta mejor   
2     4642283  Reemplazos con [[Usuario:Benjavalero/Replacer|...   
3     5338508  Reemplazos con [[Usuario:Benjavalero/Replacer|...   
4     3615366  Correcciones ortográficas con [[Usuario:Benjav...   
5      689866  Correcciones ortográficas con [[Usuario:Benjav...   

   cantComnent_Revisor  
0                    1  
2                    1  
3                    1  
4                    1  
5                    1

In [13]:
languages.head()

babel_user babel_lang babel_level
0     4502458         FR           2
1     5928200         aa           0
2       46918         ab           0
3     2050449         ab           1
4     4715583        ace           0

In [14]:
languages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35666 entries, 0 to 35665
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   babel_user   35666 non-null  int64 
 1   babel_lang   35662 non-null  object
 2   babel_level  35666 non-null  object
dtypes: int64(1), object(2)
memory usage: 836.0+ KB


In [15]:
languages.babel_level.value_counts()

N    11534
1     7589
2     6453
3     5850
4     2504
0     1092
5      644
Name: babel_level, dtype: int64

### Nos quedamos con los distintos de 0 osea solo los que tienen un nivel de idioma

In [16]:
languages_sin0=languages[(languages['babel_level']=='1')|(languages['babel_level']=='2')|(languages['babel_level']=='3')|\
                               (languages['babel_level']=='4')|(languages['babel_level']=='5')|(languages['babel_level']=='N')]

In [17]:
languages_sin0.babel_level.value_counts()

N    11534
1     7589
2     6453
3     5850
4     2504
5      644
Name: babel_level, dtype: int64

In [18]:
languages_sin0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34574 entries, 0 to 35665
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   babel_user   34574 non-null  int64 
 1   babel_lang   34572 non-null  object
 2   babel_level  34574 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.1+ MB


### Nos quedamos con los user que poseen algun idioma

In [22]:
user_languages=languages.drop_duplicates(subset=['babel_user'])
user_languages.head()

babel_user babel_lang babel_level
0     4502458         FR           2
1     5928200         aa           0
2       46918         ab           0
3     2050449         ab           1
4     4715583        ace           0

In [23]:
user_languages=user_languages[['babel_user']]
user_languages.columns=['revisor_id']
user_languages.head()

revisor_id
0     4502458
1     5928200
2       46918
3     2050449
4     4715583

In [25]:
comentario=comentario[['revisor_id','cantComnent_Revisor']]
comentario.head()

revisor_id  cantComnent_Revisor
0     1708233                    1
2     4642283                    1
3     5338508                    1
4     3615366                    1
5      689866                    1

### Ahora hacemos el merge entre los user que saben algun idioma y los usarios que hicieron algun comentario. Así nos aseguramos que los users sean solo los que saben algún idioma. 

In [32]:
final=pd.merge(user_languages,comentario,on='revisor_id',how='inner')
final.head()

revisor_id  cantComnent_Revisor
0       46918                    0
1       46918                    0
2       46918                    0
3       46918                    0
4       46918                    1

In [33]:
print(final.shape)


(824438, 2)


In [34]:
final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 824438 entries, 0 to 824437
Data columns (total 2 columns):
 #   Column               Non-Null Count   Dtype
---  ------               --------------   -----
 0   revisor_id           824438 non-null  int64
 1   cantComnent_Revisor  824438 non-null  int64
dtypes: int64(2)
memory usage: 18.9 MB


In [35]:
final['cantidad']=1
final=pd.pivot_table(final,values='cantidad',columns='cantComnent_Revisor',index='revisor_id',aggfunc=np.sum)
final.fillna(0,inplace=True)
final.head()

cantComnent_Revisor    0     1
revisor_id                    
30                   1.0  32.0
38                   2.0  10.0
311                  0.0   1.0
322                  3.0  10.0
373                  2.0   1.0

In [38]:
final.columns=['cantCommentNull','cantCommentNotNull']
final.head()

cantCommentNull  cantCommentNotNull
revisor_id                                     
30                      1.0                32.0
38                      2.0                10.0
311                     0.0                 1.0
322                     3.0                10.0
373                     2.0                 1.0

In [39]:
final['cantTotalComments']=final['cantCommentNull']+final['cantCommentNotNull']
final.head()

cantCommentNull  cantCommentNotNull  cantTotalComments
revisor_id                                                        
30                      1.0                32.0               33.0
38                      2.0                10.0               12.0
311                     0.0                 1.0                1.0
322                     3.0                10.0               13.0
373                     2.0                 1.0                3.0

In [40]:
final['tasaPedida']=final['cantCommentNotNull']/final['cantTotalComments']
final.head(10)

cantCommentNull  cantCommentNotNull  cantTotalComments  tasaPedida
revisor_id                                                                    
30                      1.0                32.0               33.0    0.969697
38                      2.0                10.0               12.0    0.833333
311                     0.0                 1.0                1.0    1.000000
322                     3.0                10.0               13.0    0.769231
373                     2.0                 1.0                3.0    0.333333
625                     3.0                51.0               54.0    0.944444
805                     0.0                33.0               33.0    1.000000
897                     1.0                14.0               15.0    0.933333
967                     0.0                 3.0                3.0    1.000000
1279                    1.0                 3.0                4.0    0.750000

### Podemos ver diferentes opciones, como los 10 usuarios cuya cantidad de comentarios son maximo

In [41]:
final.nlargest(10,'cantTotalComments')

cantCommentNull  cantCommentNotNull  cantTotalComments  tasaPedida
revisor_id                                                                    
474907                282.0             37693.0            37975.0    0.992574
58607                3372.0             32465.0            35837.0    0.905907
13349                 292.0             27871.0            28163.0    0.989632
2075                 1563.0             20918.0            22481.0    0.930475
648882              12413.0              6511.0            18924.0    0.344060
136933                 13.0             18722.0            18735.0    0.999306
2056847               745.0             16219.0            16964.0    0.956083
4002228              3443.0             12011.0            15454.0    0.777210
1628276               213.0             14071.0            14284.0    0.985088
169235                 20.0             12506.0            12526.0    0.998403